### PR TITLE
Improve check()

### LINKED ISSUE #681: Need to save metadata if thin data device is extended
Otherwise, creating a pool, doing something that causes it to extend (such as creating a filesystem) and then killing/restarting stratisd will result in stratisd failing to run.

### PR DESCRIPTION
**Pull request by agrover**
Change check to return a bool to communicate if a change has been made that
requires us to save a new version of the on-disk metadata.

Add an assert in event_on(). We will in the future possibly get events for
other DM devices in the pool, but right now we should not.

fixes #681

Signed-off-by: Andy Grover <agrover@redhat.com>

### CODE REVIEW

File: src/engine/strat_engine/thinpool/thinpool.rs [hunk 1]
```diff
@@ -314,7 +319,7 @@ impl ThinPool {
                     // A more sophisticated approach might be in order.
                     match self.extend_thinpool(dm, usage.total_data, bd_mgr) {
                         #![allow(single_match)]
-                        Ok(_) => {}
+                        Ok(_) => should_save = true,
```
Comment (mulkieran) [inline src/engine/strat_engine/thinpool/thinpool.rs]: If you are going to take this approach for ```extend_thinpool``` call, it also applies to ```extend_thinpool_meta```.

Comment (mulkieran) [inline src/engine/strat_engine/thinpool/thinpool.rs]: If you don't do the same thing for the ```Ok``` as for the ```Err```, you can remove the ```single_match``` clippy annotation.

File: src/engine/strat_engine/pool.rs [hunk 2]
```diff
@@ -128,10 +131,15 @@ impl StratPool {
         self.thin_pool.get_eventing_dev_names()
     }
 
-    pub fn event_on(&mut self, _dm_name: &DmName) -> EngineResult<()> {
-        // TODO: Just check the device that evented. Currently checks
-        // everything.
-        self.check()
+    pub fn event_on(&mut self, dm_name: &DmName) -> EngineResult<()> {
```
[the remark this answers is not in the document]
Comment (agrover) [reply src/engine/strat_engine/pool.rs]: ok

File: src/engine/strat_engine/pool.rs [hunk 3]
```diff
@@ -128,10 +131,15 @@ impl StratPool {
...
+        assert!(self.thin_pool
```
[the remark this answers is not in the document]
Comment (agrover) [reply src/engine/strat_engine/pool.rs]: ok

File: src/engine/strat_engine/thinpool/thinpool.rs (hunk 1 above)
Comment (agrover) [reply src/engine/strat_engine/thinpool/thinpool.rs]: very true! ok.

Comment (agrover) [reply src/engine/strat_engine/thinpool/thinpool.rs]: great! ok.

**Comment by agrover** [comment]
Made requested changes.

**Comment by mulkieran** [comment]
clippy wants some further changes.

**Comment by mulkieran** [comment]
MALR

**Pull request merged**